src/backend/distributed/planner/merge_planner.c

```diff
@@ -1047,22 +1051,41 @@ DeferErrorIfRoutableMergeNotSupported(Query *query, List *rangeTableList,
 							 "must be colocated", NULL, NULL);
 	}
 
-	DeferredErrorMessage *deferredError =
-		DeferErrorIfUnsupportedSubqueryPushdown(query,
-												plannerRestrictionContext);
-	if (deferredError)
-	{
-		ereport(DEBUG1, (errmsg("Sub-query is not pushable, try repartitioning")));
-		return deferredError;
-	}
+	DeferredErrorMessage *deferredError = NULL;
+
 
-	if (HasDangerousJoinUsing(query->rtable, (Node *) query->jointree))
+	/*
+	 * if the query goes to a single node ("router" in Citus' parlance),
+	 * we don't need to go through certain SQL support and colocation checks.
+	 *
+	 * For PG16+, this is required as some of the outer JOINs are converted to
+	 * "ON(true)" and filters are pushed down to the table scans. As
+	 * DeferErrorIfUnsupportedSubqueryPushdown rely on JOIN filters, it will fail to
+	 * detect the router case. However, we can still detect it by checking if
+	 * the query is a router query as the router query checks the filters on
+	 * the tables.
+	 */
+
+
+	if (!MergeSourceHasRouterSelect(query, plannerRestrictionContext))
```

Comment:
May be this way is cleaner, move `DeferErrorIfTargetHasFalseClause` to here and add?

```
if (MergeSourceHasRouterSelect)
{
                 return NULL
}
```

Reply:
I actually have a slight preference for the current state mostly to be robust for future code changes. I don't want a refactor to break this logic in any way. Is that OK for you?

Reply:
I'm fine with it, suggestion was meant for improving code-readability, yeah. but that has potential pitfalls when someone else changes the code without understanding the ramifications.